File: modules/agent/scripts/seed_documents.py
"""Upload seed document files to storage.

Run after 012_documents_seed.sql to upload actual PDF files.

Usage:
  docker compose exec agent python /app/scripts/seed_documents.py

This script uploads PDF files from seeders/documents/ to storage.
The SQL seeder (012_documents_seed.sql) creates the DB records.
"""

import asyncio
import logging
import os
import sys
from pathlib import Path

import asyncpg

# Add src to path for lib imports without triggering circular imports
sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from lib.storage import get_storage

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

SEEDS_DIR = Path(__file__).parent.parent / "seeders" / "documents"

SEED_FILES = [
    "company_proposal.pdf",
    "meeting_notes.pdf",
    "contract_draft.pdf",
    "product_spec.pdf",
    "invoice_sample.pdf",
    "individual_resume.pdf",
]


async def upload_seed_files():
    """Upload seed PDF files to storage for all tenants."""
    storage = get_storage()

    # Get all active tenants using direct asyncpg connection
    db_host = os.environ.get("POSTGRES_HOST")
    db_port = os.environ.get("POSTGRES_PORT")
    db_user = os.environ.get("POSTGRES_USER")
    db_password = os.environ.get("POSTGRES_PASSWORD")
    db_name = os.environ.get("POSTGRES_DB")

    if not all([db_host, db_port, db_user, db_password, db_name]):
        logger.warning("POSTGRES_* env vars not set, skipping seed document upload")
        return

    conn = await asyncpg.connect(
        host=db_host,
        port=int(db_port),
        user=db_user,
        password=db_password,
        database=db_name,
    )
    try:
        rows = await conn.fetch('SELECT id FROM "Tenant" WHERE status = \'active\' ORDER BY id')
        tenant_ids = [row["id"] for row in rows]
    finally:
        await conn.close()

    if not tenant_ids:
        logger.warning("No active tenants found")
        return

    uploaded_count = 0
    skipped_count = 0

    for tenant_id in tenant_ids:
        for filename in SEED_FILES:
            filepath = SEEDS_DIR / filename
            if not filepath.exists():
                logger.warning(f"File not found: {filepath}")
                continue

            storage_path = f"{tenant_id}/seed/{filename}"

            if await storage.exists(storage_path):
                skipped_count += 1
                continue

            with open(filepath, "rb") as f:
                content = f.read()

            await storage.upload(storage_path, content, "application/pdf")
            logger.info(f"Uploaded: {storage_path}")
            uploaded_count += 1

    if uploaded_count > 0:
        logger.info(f"Seed file upload complete: {uploaded_count} uploaded, {skipped_count} skipped")
    elif skipped_count > 0:
        logger.info(f"Seed files already exist ({skipped_count} files), skipping upload")


async def main():
    """Main entry point."""
    logger.info("Starting seed file upload...")
    await upload_seed_files()


if __name__ == "__main__":
    asyncio.run(main())

File: modules/agent-go/internal/etl/seed_documents.go
package etl

import (
	"context"
	"database/sql"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

// SeedFiles defines the files to upload
var SeedFiles = []string{
	// Generic seed documents
	"company_proposal.pdf",
	"meeting_notes.pdf",
	"contract_draft.pdf",
	"product_spec.pdf",
	"invoice_sample.pdf",
	// Personal documents
	"william_hubenschmidt_resume.pdf",
	"william_hubenschmidt_summary.txt",
	"william_hubenschmidt_sample_answers.txt",
	"william_hubenschmidt_coverletter_1.pdf",
	"william_hubenschmidt_coverletter_2.pdf",
}

// ContentTypes maps file extensions to MIME types
var ContentTypes = map[string]string{
	".pdf": "application/pdf",
	".txt": "text/plain",
}

// SeedDocumentsResult holds the result of seeding documents
type SeedDocumentsResult struct {
	Uploaded int
	Skipped  int
	Errors   []string
}

// SeedDocuments uploads seed document files to S3 storage
func SeedDocuments(db *sql.DB, seedsDir string) (*SeedDocumentsResult, error) {
	result := &SeedDocumentsResult{}

	// Get S3 configuration from environment
	bucket := os.Getenv("S3_BUCKET")
	region := os.Getenv("AWS_REGION")
	endpoint := os.Getenv("S3_ENDPOINT")
	accessKey := os.Getenv("AWS_ACCESS_KEY_ID")
	secretKey := os.Getenv("AWS_SECRET_ACCESS_KEY")

	if bucket == "" {
		return nil, fmt.Errorf("S3_BUCKET environment variable not set")
	}

	// Create S3 client
	ctx := context.Background()

	var cfg aws.Config
	var err error

	if endpoint != "" {
		// Custom endpoint (e.g., MinIO, LocalStack)
		customResolver := aws.EndpointResolverWithOptionsFunc(func(service, region string, options ...interface{}) (aws.Endpoint, error) {
			return aws.Endpoint{
				URL:               endpoint,
				HostnameImmutable: true,
			}, nil
		})

		cfg, err = config.LoadDefaultConfig(ctx,
			config.WithRegion(region),
			config.WithEndpointResolverWithOptions(customResolver),
			config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(accessKey, secretKey, "")),
		)
	} else {
		cfg, err = config.LoadDefaultConfig(ctx, config.WithRegion(region))
	}

	if err != nil {
		return nil, fmt.Errorf("failed to load AWS config: %w", err)
	}

	client := s3.NewFromConfig(cfg)

	// Get active tenants
	rows, err := db.Query(`SELECT id FROM "Tenant" WHERE status = 'active' ORDER BY id`)
	if err != nil {
		return nil, fmt.Errorf("failed to query tenants: %w", err)
	}
	defer rows.Close()

	var tenantIDs []int64
	for rows.Next() {
		var id int64
		if err := rows.Scan(&id); err != nil {
			continue
		}
		tenantIDs = append(tenantIDs, id)
	}

	if len(tenantIDs) == 0 {
		fmt.Println("No active tenants found")
		return result, nil
	}

	fmt.Printf("Found %d active tenant(s)\n", len(tenantIDs))

	// Upload seed files for each tenant
	for _, tenantID := range tenantIDs {
		for _, filename := range SeedFiles {
			localPath := filepath.Join(seedsDir, filename)
			if _, err := os.Stat(localPath); os.IsNotExist(err) {
				fmt.Printf("Warning: File not found: %s\n", localPath)
				continue
			}

			storagePath := fmt.Sprintf("%d/seed/%s", tenantID, filename)

			// Check if file already exists
			_, err := client.HeadObject(ctx, &s3.HeadObjectInput{
				Bucket: aws.String(bucket),
				Key:    aws.String(storagePath),
			})
			if err == nil {
				result.Skipped++
				continue
			}

			// Read file content
			content, err := os.ReadFile(localPath)
			if err != nil {
				result.Errors = append(result.Errors, fmt.Sprintf("failed to read %s: %v", filename, err))
				continue
			}

			// Determine content type
			ext := strings.ToLower(filepath.Ext(filename))
			contentType := ContentTypes[ext]
			if contentType == "" {
				contentType = "application/octet-stream"
			}

			// Upload to S3
			_, err = client.PutObject(ctx, &s3.PutObjectInput{
				Bucket:      aws.String(bucket),
				Key:         aws.String(storagePath),
				Body:        strings.NewReader(string(content)),
				ContentType: aws.String(contentType),
			})
			if err != nil {
				result.Errors = append(result.Errors, fmt.Sprintf("failed to upload %s: %v", storagePath, err))
				continue
			}

			fmt.Printf("Uploaded: %s\n", storagePath)
			result.Uploaded++
		}
	}

	if result.Uploaded > 0 {
		fmt.Printf("Seed file upload complete: %d uploaded, %d skipped\n", result.Uploaded, result.Skipped)
	} else if result.Skipped > 0 {
		fmt.Printf("Seed files already exist (%d files), skipping upload\n", result.Skipped)
	}

	return result, nil
}

// GetSeedsDir returns the seed documents directory path
func GetSeedsDir() string {
	// Docker path
	if _, err := os.Stat("/app/seeders/documents"); err == nil {
		return "/app/seeders/documents"
	}

	// Local development
	candidates := []string{
		"seeders/documents",
		"../seeders/documents",
		"../../seeders/documents",
	}

	for _, dir := range candidates {
		if _, err := os.Stat(dir); err == nil {
			return dir
		}
	}

	return "seeders/documents"
}
